Replace JSONPath with `jq` in `jsondocck`

fluiderson opened the pull request:
This is an ongoing effort to close https://github.com/rust-lang/rust/issues/142479. I've decided to take a hard path and rewrite `rustdoc-types`'s entire testsuite. I don't think it'll be a good idea to construct a JSONPath-`jq` transpiler because it'll look really awkward for new contributors, JSONPath's limitations + [`jaq`'s own quirks](https://github.com/01mf02/jaq/blob/8c5131b4060d95856cc347e98baaf270b5bd43a8/README.md#miscellaneous) may create a lot of questions that even current maintainers would've no answers for without digging into `jsondocck`'s internals.

I expect `jsondocck` to have just 2 directives: `jq` and `arg`. Thanks to `jq`'s flexibility, these 2 can do everything that the current directive set can and much more (actually, too much, I left some comments to opt out some dependencies when it'll be possible on `jaq`'s side). I'm trying to actively utilize `jq`'s features instead of just blindly rewriting existing assertions, so there should be more deletions than additions.

When I'll be done, I'll try to rebase and split my work to make it more easier to review. So far `jsondocck` is already functional, so you can check out how it may look like and give some feedback if you want. I'll commit updated tests folder by folder.

fluiderson committed:
update the edition

fluiderson committed:
remove the version field

it's unnecessary for private crates

fluiderson committed:
remove an unused variable

the deleted lines are duplicates of the lines 36 & 37

fluiderson committed:
appease clippy

fluiderson committed:
unify message style

fluiderson committed:
add "input"/"output" to make it more clear what's what

fluiderson committed:
unify message style

fluiderson committed:
optimize the config parsing

fluiderson committed:
unicode is enabled by default

https://docs.rs/regex/1/regex/struct.RegexBuilder.html#method.unicode

fluiderson committed:
disable the verbose mode for `DEPRECATED_LINE_PATTERN`

there are no insignificant whitespaces or comments, it's a single line

fluiderson committed:
modify `lineno` instead of shadowing

fluiderson committed:
unify the error message style, skip directives collection

- directives were collected before processing, but there seems to be nothing preventing them from checking right away
- the unified error message style follows the style of short diagnostics from the compiler (almost the same was used on the deleted line 39)

fluiderson committed:
add a new pattern to catch ignored directives on mixed lines

fluiderson committed:
more error message style unifying

fluiderson committed:
remove unnecessary hashes

fluiderson committed:
`attrs/` is done

fluiderson committed:
`enums/discriminant/` is done

fluiderson committed:
`enums/` is done

fluiderson committed:
`fn_pointer/` is done

fluiderson committed:
`fns/` is done

bors commented:
:umbrella: The latest upstream changes (presumably #143337) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

fmease commented:
(cc @aDotInTheVoid)

aDotInTheVoid commented:
Sorry for taking so long to review this. Thanks for working on it, especially the repetitive test migration.


>  I don't think it'll be a good idea to construct a JSONPath-`jq` transpiler because it'll look really awkward for new contributors

What do you mean by this? The idea was to use the transpiler to migrate all the assertions in the tests, leaving them in jq. New contributors should never need to know it existed.

> I expect `jsondocck` to have just 2 directives: `jq` and `arg`

I think this is a quite bad idea. It means we loose any sort of good error reporting. With this PR at the moment, we can only say that a check failed, but not why (in the case that the query matched the wrong value). Having this is really important I think.

fluiderson committed:
simplify the `LINE_PATTERN` regex

as for the start and end anchors (`^` & `$`), they're unnecessary since we apply this regex to separate lines, not whole text

fluiderson committed:
rewrite `jsondocck` from JSONPath to `jq`

i tried to keep the original code structure, but it's an almost entire rewrite. thankfully, there are more deletions than insertions, i expect jq to be flexible enough to supersede almost the entire previous set of directives. no tests yet, but it seems to be working. jaq's guts are a bit arcane to me, but it seems to be only related to its setup, a resulted API turned out to be pretty straightforward

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16207589968/job/45761306371) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45761306371) [(plain)](https://github.com/rust-lang/rust/commit/97d350d79806f1987b02565e055bbda7c8138b2a/checks/45761306371/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
All checks passed!
checking python file formatting
28 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:20
  local time: Thu Jul 10 22:52:06 UTC 2025
  network time: Thu, 10 Jul 2025 22:52:06 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #143810) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

aDotInTheVoid commented:
> I see. I've added 2 directives `eq` & `ne`. They act like `is` & `!is` and report what values they received in case of mismatch. You can see how it looks in the `attrs/` tests directory. So, in total, `arg`, `jq`, `eq`, and `ne`. Would that be enough?

I don't think this PR needs to change the directives that exist at all to be honest. Why do you think that's the right approach?

What's the benefit of renaming `is`/`!is` to `eq`/`ne` (and `set` to `arg`)? It seems like it's churn for no gain, but maybe I'm missing something.

I think the best approach would be to keep all the existing directives, and not do a general `jq` directive. (That behavior could be with something like `//@ is '.some.jq.query' true`, but I don't think this should be encouraged). But this is based just on thinking about it, and you've written a load of tests in jq. I'm curious which way you think makes sense, and *why*.

aDotInTheVoid commented:
>  For nonexistent paths, `jq` returns `null`, but [if an existing path contains `null`, you'll get `null`, too](https://play.jqlang.org/s/tnSxBEc798lgH1g)

Wah, that's very unfortunate. We should probably ban `//@ is .some.path null`, as that's super footgunny. But 

> - `is [.], [1, 2, 3]` instead of `ismany`

Dylan-DPC commented:
@fluiderson any updates on this? thanks

fluiderson commented:
No updates. I'm busy with other stuff right now. I'll probably return to this PR later this year.
